lib/admin.rb

```diff
@@ -0,0 +1,149 @@
+  end
+  # param: reservation number
+  # returns - new reservation instance
+  def request_reservation_within_block(id, start_date, end_date)
+    #split this up into smaller methods possibly?
+    check_date(start_date, end_date)
+    check_id(id)
+
+    block_reservation = find_reservation_by_id(id)
+    # should I postfix these conditionals if it ends up exceeding the lines? -->
+    if Date.parse(start_date) != block_reservation.start_date
+      raise StandardError, "start date must match block start"
+    end
+    if Date.parse(end_date) != block_reservation.end_date
+      raise StandardError, "end date must match block end"
+    end
+
+    id = block_reservation.reservations.length + 1
+    room = block_reservation.rooms_available.sample
+      raise ArgumentError, "no rooms available" if room.nil?
+
+    new_reservation = Reservation.new(id, room, start_date, end_date,
+      room_cost: 150)
+
+    add_reservation(new_reservation, reservation_location: block_reservation.reservations)
+    # deletes room number from available_rooms
+    block_reservation.rooms_available.delete(new_reservation.room)
+
+    return new_reservation
+  end
+
+  def find_reservation_by_id(id)
+    check_id(id)
+
+    specific_reservation = @reservations.find { |reservation|
+      reservation.id == id }
+    return specific_reservation
+  end
+  #param - date
+  #returns - array of reservations within that date
+  def reservations_by_date(date)
+    specific_reservations = @reservations.find_all do |reservation|
+      (reservation.start_date..reservation.end_date).cover?(Date.parse(date))
+    end
+
+    return specific_reservations
+  end
+
+  def reservations_by_date_range(trip_start, trip_end)
+      trip_start = Date.parse(trip_start)
+      trip_end = Date.parse(trip_end)
+
+
+    @reservations.find_all do |reservation|
+      unless reservation.end_date == trip_start
+        (reservation.start_date..reservation.end_date).cover?(trip_start) ||
+        (reservation.start_date..reservation.end_date).cover?(trip_end)
+      end
+    end
+  end
+
+  def reservation_cost(reservation)
+    # https://stackoverflow.com/questions/4502245/how-can-i-find-the-number-of-days-between-two-date-objects-in-ruby
+    start_date = reservation.start_date
+    end_date = reservation.end_date
+    total_nights = end_date.mjd - start_date.mjd
+
+    cost = total_nights * reservation.room_cost
+
+    return cost
+  end
+
+  def available_rooms(trip_start, trip_end,
+    all_rooms: [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20] )
+    specific_reservations = reservations_by_date_range(trip_start, trip_end)
+
+    unavailable_rooms = []
+
+    specific_reservations.each do |reservation|
+      unavailable_rooms << reservation.room
+    end
+
+    unavailable_rooms.flatten!
+    available_rooms = all_rooms - unavailable_rooms
+    return available_rooms
+  end
+
+  private
+
+  def check_date(start_date, end_date)
+    regex = /^\d{4}-\d{1,2}-\d{1,2}$/
```

Comment:
Is this strictly speaking needed?

Reply:
It sends an error back if the person running the admin accidentally puts in the wrong date format.  To use date format it needs an exact, so this helps account if an accidental typo was entered.